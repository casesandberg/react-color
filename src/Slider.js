export default from './components/slider/Slider'
