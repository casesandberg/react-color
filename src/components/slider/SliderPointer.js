'use strict' /* @flow */

import React from 'react'
import ReactCSS from 'reactcss'
import shallowCompare from 'react-addons-shallow-compare'

export class SliderPointer extends ReactCSS.Component {
  shouldComponentUpdate = shallowCompare.bind(this, this, arguments[0], arguments[1])

  classes(): any {
    return {
      'default': {
        picker: {
          width: '14px',
          height: '14px',
          borderRadius: '6px',
          transform: 'translate(-7px, -1px)',
          backgroundColor: 'rgb(248, 248, 248)',
          boxShadow: '0px 1px 4px 0 rgba(0, 0, 0, 0.37)',
        },
      },
    }
  }

  render(): any {
    return (
      <div is="picker"></div>
    )
  }
}

export default SliderPointer
