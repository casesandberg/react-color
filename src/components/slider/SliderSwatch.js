'use strict' /* @flow */

import React from 'react'
import ReactCSS from 'reactcss'
import shallowCompare from 'react-addons-shallow-compare'

export class SliderSwatch extends ReactCSS.Component {
  shouldComponentUpdate = shallowCompare.bind(this, this, arguments[0], arguments[1])

  classes(): any {
    return {
      'default': {
        swatch: {
          height: '12px',
          background: 'hsl(' + this.props.hsl.h + ', 50%, ' + (this.props.offset * 100) + '%)',
          cursor: 'pointer',
        },
      },
      'first': {
        swatch: {
          borderRadius: '2px 0px 0px 2px',
        },
      },
      'last': {
        swatch: {
          borderRadius: '0px 2px 2px 0px',
        },
      },
      active: {
        swatch: {
          transform: 'scaleY(1.8)',
          borderRadius: '3.6px/2px',
        },
      },
    }
  }

  handleClick = () => {
    this.props.onClick({
      h: this.props.hsl.h,
      s: .5,
      l: this.props.offset,
      source: 'hsl',
    })
  }

  render(): any {
    return (
      <div is="swatch" ref="swatch" onClick={ this.handleClick } />
    )
  }

}

export default SliderSwatch
