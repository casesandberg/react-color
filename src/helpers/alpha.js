export const calculateChange = (e, skip, props, container) => {
  e.preventDefault()
  const containerWidth = container.clientWidth
  const containerHeight = container.clientHeight
  const x = typeof e.pageX === 'number' ? e.pageX : e.touches[0].pageX
  const y = typeof e.pageY === 'number' ? e.pageY : e.touches[0].pageY
  const left = x - (container.getBoundingClientRect().left + window.pageXOffset)
  const top = y - (container.getBoundingClientRect().top + window.pageYOffset)

  if (props.direction === 'vertical') {
    let a
    if (top < 0) {
      a = 0
    } else if (top > containerHeight) {
      a = 1
    } else {
      a = Math.round((top * 100) / containerHeight) / 100
    }

    if (props.hsl.a !== a) {
      return {
        h: props.hsl.h,
        s: props.hsl.s,
        l: props.hsl.l,
        a,
        source: 'rgb',
      }
    }
  } else {
    let a
    if (left < 0) {
      a = 0
    } else if (left > containerWidth) {
      a = 1
    } else {
      a = Math.round((left * 100) / containerWidth) / 100
    }

    if (props.a !== a) {
      return {
        h: props.hsl.h,
        s: props.hsl.s,
        l: props.hsl.l,
        a,
        source: 'rgb',
      }
    }
  }
  return null
}
