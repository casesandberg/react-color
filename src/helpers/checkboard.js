const checkboardCache = {}

export const render = (c1, c2, size, serverCanvas) => {
  if (typeof document === 'undefined' && !serverCanvas) {
    return null
  }
  const canvas = serverCanvas ? new serverCanvas() : document.createElement('canvas')
  canvas.width = size * 2
  canvas.height = size * 2
  const ctx = canvas.getContext('2d')
  if (!ctx) {
    return null
  } // If no context can be found, return early.
  ctx.fillStyle = c1
  ctx.fillRect(0, 0, canvas.width, canvas.height)
  ctx.fillStyle = c2
  ctx.fillRect(0, 0, size, size)
  ctx.translate(size, size)
  ctx.fillRect(0, 0, size, size)
  return canvas.toDataURL()
}

export const get = (c1, c2, size, serverCanvas) => {
  const key = `${ c1 }-${ c2 }-${ size }${ serverCanvas ? '-server' : '' }`

  if (checkboardCache[key]) {
    return checkboardCache[key]
  }

  const checkboard = render(c1, c2, size, serverCanvas)
  checkboardCache[key] = checkboard
  return checkboard
}
