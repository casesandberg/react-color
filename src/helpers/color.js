import each from 'lodash/each'
import tinycolor from 'tinycolor2'

export default {
  simpleCheckForValidColor(data) {
    const keysToCheck = ['r', 'g', 'b', 'a', 'h', 's', 'l', 'v']
    let checked = 0
    let passed = 0
    each(keysToCheck, (letter) => {
      if (data[letter]) {
        checked += 1
        if (!isNaN(data[letter])) {
          passed += 1
        }
        if (letter === 's' || letter === 'l') {
          const percentPatt = /^\d+%$/
          if (percentPatt.test(data[letter])) {
            passed += 1
          }
        }
      }
    })
    return (checked === passed) ? data : false
  },

  toState(data, oldHue) {
    const color = data.hex ? tinycolor(data.hex) : tinycolor(data)
    const hsl = color.toHsl()
    const hsv = color.toHsv()
    const rgb = color.toRgb()
    const hex = color.toHex()
    if (hsl.s === 0) {
      hsl.h = oldHue || 0
      hsv.h = oldHue || 0
    }
    const transparent = hex === '000000' && rgb.a === 0

    return {
      hsl,
      hex: transparent ? 'transparent' : `#${ hex }`,
      rgb,
      hsv,
      oldHue: data.h || oldHue || hsl.h,
      source: data.source,
    }
  },

  isValidHex(hex) {
    return tinycolor(hex).isValid()
  },

  getContrastingColor(data) {
    if (!data) {
      return '#fff'
    }
    const col = this.toState(data)
    if (col.hex === 'transparent') {
      return 'rgba(0,0,0,0.4)'
    }
    const yiq = ((col.rgb.r * 299) + (col.rgb.g * 587) + (col.rgb.b * 114)) / 1000
    return (yiq >= 128) ? '#000' : '#fff'
  },
}

export const red = {
  hsl: { a: 1, h: 0, l: 0.5, s: 1 },
  hex: '#ff0000',
  rgb: { r: 255, g: 0, b: 0, a: 1 },
  hsv: { h: 0, s: 1, v: 1, a: 1 },
}
