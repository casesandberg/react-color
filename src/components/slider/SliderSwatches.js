import React from 'react'
import reactCSS from 'reactcss'

import SliderSwatch from './SliderSwatch'

export const SliderSwatches = ({ onClick, hsl }) => {
  const styles = reactCSS({
    'default': {
      swatches: {
        marginTop: '20px',
      },
      swatch: {
        boxSizing: 'border-box',
        width: '20%',
        paddingRight: '1px',
        float: 'left',
      },
      clear: {
        clear: 'both',
      },
    },
  })

  // Acceptible difference in floating point equality
  const epsilon = 0.1

  return (
    <div style={ styles.swatches }>
      <div style={ styles.swatch }>
        <SliderSwatch
          hsl={ hsl }
          offset=".80"
          active={ Math.abs(hsl.l - 0.80) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon }
          onClick={ onClick }
          first
        />
      </div>
      <div style={ styles.swatch }>
        <SliderSwatch
          hsl={ hsl }
          offset=".65"
          active={ Math.abs(hsl.l - 0.65) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon }
          onClick={ onClick }
        />
      </div>
      <div style={ styles.swatch }>
        <SliderSwatch
          hsl={ hsl }
          offset=".50"
          active={ Math.abs(hsl.l - 0.50) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon }
          onClick={ onClick }
        />
      </div>
      <div style={ styles.swatch }>
        <SliderSwatch
          hsl={ hsl }
          offset=".35"
          active={ Math.abs(hsl.l - 0.35) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon }
          onClick={ onClick }
        />
      </div>
      <div style={ styles.swatch }>
        <SliderSwatch
          hsl={ hsl }
          offset=".20"
          active={ Math.abs(hsl.l - 0.20) < epsilon
            && Math.abs(hsl.s - 0.50) < epsilon }
          onClick={ onClick }
          last
        />
      </div>
      <div style={ styles.clear } />
    </div>
  )
}

export default SliderSwatches
