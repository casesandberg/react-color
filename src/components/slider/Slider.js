import React from 'react'
import reactCSS from 'reactcss'

import { ColorWrap, Hue } from '../common'
import SliderSwatches from './SliderSwatches'
import SliderPointer from './SliderPointer'

export const Slider = ({ hsl, onChange, pointer, className = '' }) => {
  const styles = reactCSS({
    'default': {
      hue: {
        height: '12px',
        position: 'relative',
      },
      Hue: {
        radius: '2px',
      },
    },
  })

  return (
    <div className={ `slider-picker ${ className }` }>
      <div style={ styles.hue }>
        <Hue
          style={ styles.Hue }
          hsl={ hsl }
          pointer={ pointer }
          onChange={ onChange }
        />
      </div>
      <div style={ styles.swatches }>
        <SliderSwatches hsl={ hsl } onClick={ onChange } />
      </div>
    </div>
  )
}

Slider.defaultProps = {
  pointer: SliderPointer,
}

export default ColorWrap(Slider)
