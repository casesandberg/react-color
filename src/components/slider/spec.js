/* global test, expect */

import React from 'react'
import renderer from 'react-test-renderer'
import { red } from '../../helpers/color'

import Slider from './Slider'
import SliderPointer from './SliderPointer'
import SliderSwatch from './SliderSwatch'
import SliderSwatches from './SliderSwatches'

test('Slider renders correctly', () => {
  const tree = renderer.create(
    <Slider { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SliderPointer renders correctly', () => {
  const tree = renderer.create(
    <SliderPointer />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SliderSwatch renders correctly', () => {
  const tree = renderer.create(
    <SliderSwatch { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SliderSwatches renders correctly', () => {
  const tree = renderer.create(
    <SliderSwatches { ...red } />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
