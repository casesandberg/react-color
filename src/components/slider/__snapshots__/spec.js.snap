// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders correctly 1`] = `
<div
  className="slider-picker "
>
  <div
    style={
      Object {
        "height": "12px",
        "position": "relative",
      }
    }
  >
    <div
      style={
        Object {
          "MozBorderRadius": undefined,
          "MozBoxShadow": undefined,
          "OBorderRadius": undefined,
          "OBoxShadow": undefined,
          "WebkitBorderRadius": undefined,
          "WebkitBoxShadow": undefined,
          "borderRadius": undefined,
          "bottom": "0px",
          "boxShadow": undefined,
          "left": "0px",
          "msBorderRadius": undefined,
          "msBoxShadow": undefined,
          "position": "absolute",
          "right": "0px",
          "top": "0px",
        }
      }
    >
      <div
        className="hue-horizontal"
        onMouseDown={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "height": "100%",
            "padding": "0 2px",
            "position": "relative",
          }
        }
      >
        <style>
          
                      .hue-horizontal {
                        background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                          33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                        background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                          17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      }
          
                      .hue-vertical {
                        background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                          #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                        background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                          #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
                      }
                    
        </style>
        <div
          style={
            Object {
              "left": "69.44444444444443%",
              "position": "absolute",
            }
          }
        >
          <div
            style={
              Object {
                "MozBorderRadius": "6px",
                "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "MozTransform": "translate(-7px, -1px)",
                "OBorderRadius": "6px",
                "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "OTransform": "translate(-7px, -1px)",
                "WebkitBorderRadius": "6px",
                "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "WebkitTransform": "translate(-7px, -1px)",
                "backgroundColor": "rgb(248, 248, 248)",
                "borderRadius": "6px",
                "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "height": "14px",
                "msBorderRadius": "6px",
                "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
                "msTransform": "translate(-7px, -1px)",
                "transform": "translate(-7px, -1px)",
                "width": "14px",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    style={undefined}
  >
    <div
      style={
        Object {
          "marginTop": "20px",
        }
      }
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            Object {
              "MozBorderRadius": "2px 0 0 2px",
              "OBorderRadius": "2px 0 0 2px",
              "WebkitBorderRadius": "2px 0 0 2px",
              "background": "hsl(249.99999999999994, 50%, 80%)",
              "borderRadius": "2px 0 0 2px",
              "cursor": "pointer",
              "height": "12px",
              "msBorderRadius": "2px 0 0 2px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            Object {
              "background": "hsl(249.99999999999994, 50%, 65%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            Object {
              "background": "hsl(249.99999999999994, 50%, 50%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            Object {
              "background": "hsl(249.99999999999994, 50%, 35%)",
              "cursor": "pointer",
              "height": "12px",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "float": "left",
            "paddingRight": "1px",
            "width": "20%",
          }
        }
      >
        <div
          onClick={[Function]}
          style={
            Object {
              "MozBorderRadius": "0 2px 2px 0",
              "MozTransform": "scaleY(1.8)",
              "OBorderRadius": "0 2px 2px 0",
              "OTransform": "scaleY(1.8)",
              "WebkitBorderRadius": "0 2px 2px 0",
              "WebkitTransform": "scaleY(1.8)",
              "background": "hsl(249.99999999999994, 50%, 20%)",
              "borderRadius": "0 2px 2px 0",
              "cursor": "pointer",
              "height": "12px",
              "msBorderRadius": "0 2px 2px 0",
              "msTransform": "scaleY(1.8)",
              "transform": "scaleY(1.8)",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "clear": "both",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`SliderPointer renders correctly 1`] = `
<div
  style={
    Object {
      "MozBorderRadius": "6px",
      "MozBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "MozTransform": "translate(-7px, -1px)",
      "OBorderRadius": "6px",
      "OBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "OTransform": "translate(-7px, -1px)",
      "WebkitBorderRadius": "6px",
      "WebkitBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "WebkitTransform": "translate(-7px, -1px)",
      "backgroundColor": "rgb(248, 248, 248)",
      "borderRadius": "6px",
      "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "height": "14px",
      "msBorderRadius": "6px",
      "msBoxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "msTransform": "translate(-7px, -1px)",
      "transform": "translate(-7px, -1px)",
      "width": "14px",
    }
  }
/>
`;

exports[`SliderSwatch renders correctly 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "background": "hsl(0, 50%, NaN%)",
      "cursor": "pointer",
      "height": "12px",
    }
  }
/>
`;

exports[`SliderSwatches renders correctly 1`] = `
<div
  style={
    Object {
      "marginTop": "20px",
    }
  }
>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "MozBorderRadius": "2px 0 0 2px",
          "OBorderRadius": "2px 0 0 2px",
          "WebkitBorderRadius": "2px 0 0 2px",
          "background": "hsl(0, 50%, 80%)",
          "borderRadius": "2px 0 0 2px",
          "cursor": "pointer",
          "height": "12px",
          "msBorderRadius": "2px 0 0 2px",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "background": "hsl(0, 50%, 65%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "background": "hsl(0, 50%, 50%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "background": "hsl(0, 50%, 35%)",
          "cursor": "pointer",
          "height": "12px",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "float": "left",
        "paddingRight": "1px",
        "width": "20%",
      }
    }
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "MozBorderRadius": "0 2px 2px 0",
          "OBorderRadius": "0 2px 2px 0",
          "WebkitBorderRadius": "0 2px 2px 0",
          "background": "hsl(0, 50%, 20%)",
          "borderRadius": "0 2px 2px 0",
          "cursor": "pointer",
          "height": "12px",
          "msBorderRadius": "0 2px 2px 0",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
</div>
`;
